Change URL matching code to restore compatibility with old pprof (#276)

ErwanLegrand opened the pull request:
Here are the changes I have made to restore compatibility with the old pprof. I have left `strings.Contains(url.Path, "/debug/pprof/")` in the condition, together with the two `HasSuffix()` tests I have added. Leaving it seemed like the safe thing to do.

ErwanLegrand commented:
I have signed it!

(I would have thought I had signed this already. I'm pretty sure I made some minor contribution to Go in the past.)

aalexand commented:
Thank you! There is symbolz_test.go with tests, could you update it with some test cases that fail before your change and pass after?

ErwanLegrand commented:
I did not even notice the test file! I'll update it right away.

ErwanLegrand commented:
I can't run the tests locally. I get the following errors:

> + go test -v github.com/ErwanLegrand/pprof/fuzz github.com/ErwanLegrand/pprof/internal/binutils github.com/ErwanLegrand/pprof/internal/driver github.com/ErwanLegrand/pprof/internal/elfexec github.com/ErwanLegrand/pprof/internal/graph github.com/ErwanLegrand/pprof/internal/measurement github.com/ErwanLegrand/pprof/internal/report github.com/ErwanLegrand/pprof/internal/symbolizer github.com/ErwanLegrand/pprof/internal/symbolz github.com/ErwanLegrand/pprof/profile
> internal/binutils/binutils.go:29:2: use of internal package not allowed
> internal/binutils/addr2liner.go:26:2: use of internal package not allowed
> internal/driver/cli.go:22:2: use of internal package not allowed
> internal/driver/flamegraph.go:23:2: use of internal package not allowed
> internal/driver/driver_focus.go:23:2: use of internal package not allowed
> internal/driver/cli.go:23:2: use of internal package not allowed
> internal/driver/commands.go:30:2: use of internal package not allowed
> internal/driver/options.go:27:2: use of internal package not allowed
> internal/graph/dotgraph.go:24:2: use of internal package not allowed
> internal/report/report.go:30:2: use of internal package not allowed
> internal/report/report.go:31:2: use of internal package not allowed
> internal/report/report.go:32:2: use of internal package not allowed
> internal/symbolizer/symbolizer.go:29:2: use of internal package not allowed
> internal/symbolizer/symbolizer.go:30:2: use of internal package not allowed
> internal/symbolizer/symbolizer.go:31:2: use of internal package not allowed
> internal/symbolz/symbolz.go:29:2: use of internal package not allowed

I'll update the pull request with the new tests anyway.

aalexand commented:
@ErwanLegrand see #279

ErwanLegrand commented:
@aalexand now it works. Thanks!

ErwanLegrand commented:
Looking at the tests in `symbolz_test.go`, this looks like I could have taken away the `strings.Contains(url.Path, "/debug/pprof/")` test, after all. What do you think, @aalexand ?

aalexand commented:
With Go, there are also paths like `/debug/pprof/mutex`, see https://golang.org/pkg/net/http/pprof/, so no. Adding a test for that, or more generically for `/debug/pprof/foo` would be good to be explicit about why that check is still required. Can you add that please?

ErwanLegrand commented:
Good catch! I have added tests for these and for the URIs listed in https://gperftools.github.io/gperftools/pprof_remote_servers.html

I have also added support for the three listed cases which were not handled. I believe that all the cases are now covered.

aalexand commented on internal/symbolz/symbolz.go:
```diff
@@ -63,10 +63,29 @@ func Symbolize(p *profile.Profile, force bool, sources plugin.MappingSources, sy
+// Check whether path ends with one of the suffixes listed in
+// pprof_remote_servers.html from the gperftools distribution
+func hasLegacySuffix(path string) bool {
+	var suffixes = [...]string{
```
You shouldn't need the `...`.

ErwanLegrand replied:
It seemed logical to me to make this a fixed size array rather than a slice. I guess this does not matter much in practice, though.

aalexand commented on internal/symbolz/symbolz.go:
```diff
@@ -63,10 +63,29 @@ func Symbolize(p *profile.Profile, force bool, sources plugin.MappingSources, sy
+// Check whether path ends with one of the suffixes listed in
+// pprof_remote_servers.html from the gperftools distribution
+func hasLegacySuffix(path string) bool {
+	var suffixes = [...]string{
```
Can be const:
```
const suffixes := []string{
  ...
}
```

ErwanLegrand replied:
It seems arrays and slices cannot be made const:
```
internal/symbolz/symbolz.go:69:8: const initializer [5]string literal is not a constant
```

ErwanLegrand replied:
Const arrays (and slices) are discussed here: https://github.com/golang/go/issues/6386

aalexand commented on internal/symbolz/symbolz.go:
```diff
@@ -63,10 +63,29 @@ func Symbolize(p *profile.Profile, force bool, sources plugin.MappingSources, sy
 
+// Check whether path ends with one of the suffixes listed in
+// pprof_remote_servers.html from the gperftools distribution
+func hasLegacySuffix(path string) bool {
```
I am not sure "legacy" is the best name. The remote serving is a feature supported by gperftools, right? Is that really "legacy"? Maybe it should be called "hasGperftoolsSuffix"?

ErwanLegrand replied:
The gperftools devs themselves call their version of pprof legacy and recommend switching to this one. Yet, `hasGperftoolsSuffix` would certainly be clearer.

aalexand commented on internal/symbolz/symbolz.go:
```diff
@@ -63,10 +63,29 @@ func Symbolize(p *profile.Profile, force bool, sources plugin.MappingSources, sy
 func symbolz(source string) string {
 	if url, err := url.Parse(source); err == nil && url.Host != "" {
-		if strings.Contains(url.Path, "/debug/pprof/") {
+		// All paths in the net/http/pprof Go package start with /debug/pprof/
```
Not necessarily "start" - may rather contain "/debug/pprof" because the code and #249.

aalexand commented on internal/symbolz/symbolz.go:
```diff
@@ -65,7 +65,7 @@ func Symbolize(p *profile.Profile, force bool, sources plugin.MappingSources, sy
 // pprof_remote_servers.html from the gperftools distribution
-func hasLegacySuffix(path string) bool {
+func hasGperftoolsSuffix(path string) bool {
 	var suffixes = [...]string{
```
Using `var suffixes = [...]string{` is not quite idiomatic. `suffixes := []string{` is more so.

aalexand commented on CONTRIBUTING.md:
```diff
@@ -11,6 +11,45 @@ through the issue tracker with your idea so that we can help out and possibly
 guide you. Coordinating up front makes it much easier to avoid frustration later
 on.
 
+# Development
```
This shouldn't be in this PR?

ErwanLegrand replied:
Argh! :-(

Probably I have rebased against the wrong branch at some point.

aalexand replied:
Git can be hard!

ErwanLegrand replied:
Even more so when you have to switch between different git workflows!
